### PR TITLE
[heft] Load configuration files in plugins that use them instead of in a central place.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: common/changes/@rushstack/heft/ianc-heft-config-loading_2020-09-12-01-44.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft",
+      "comment": "Rename the \"outFolderPath\" and \"emitFolderPathForJest\" properties in typescript.json to \"outFolderName\" and \"emitFolderNameForJest\"",
```
Comment (octogonz) [inline common/changes/@rushstack/heft/ianc-heft-config-loading_2020-09-12-01-44.json:5]:
```suggestion
      "comment": "(BREAKING CHANGE) Rename the \"outFolderPath\" and \"emitFolderPathForJest\" properties in typescript.json to \"outFolderName\" and \"emitFolderNameForJest\"",
```

Breaking changes need to be called out very clearly, so people aren't confused when they upgrade.

File: apps/heft/src/stages/CleanStage.ts [hunk 2]
```diff
@@ -36,16 +37,43 @@ export interface ICleanStageOptions {
...
 export class CleanStage extends StageBase<CleanStageHooks, ICleanStageProperties, ICleanStageOptions> {
   public constructor(heftConfiguration: HeftConfiguration, loggingManager: LoggingManager) {
     super(heftConfiguration, loggingManager, CleanStageHooks);
   }
 
-  protected getDefaultStageProperties(options: ICleanStageOptions): ICleanStageProperties {
+  protected async getDefaultStagePropertiesAsync(
+    options: ICleanStageOptions
+  ): Promise<ICleanStageProperties> {
+    const cleanConfigurationFileLoader: ConfigurationFile<ICleanConfigurationJson> = new ConfigurationFile<
```
Comment (octogonz) [inline apps/heft/src/stages/CleanStage.ts:52]: Can we separate the "declaration of a config file" from the "loading of the config file"?  For example, suppose the code here reduced to something like:

```ts
const cleanConfigurationFileLoader: ConfigurationFile<ICleanConfigurationJson> 
  = HeftConfigFiles.clean.loadAsync(myPath);
```

Then if I'm trying to find the code that loads `clean.schema.json`, I can go to `HeftConfigFiles.ts`  and all the declarations are nicely arranged there.

File: apps/heft/src/stages/CleanStage.ts [hunk 3]
```diff
@@ -36,16 +37,43 @@ export interface ICleanStageOptions {
...
+      }
+    });
+
+    let cleanConfigurationFile: ICleanConfigurationJson | undefined = undefined;
+    try {
+      cleanConfigurationFile = await cleanConfigurationFileLoader.loadConfigurationFileAsync(
+        path.resolve(this.heftConfiguration.buildFolder, '.heft', 'clean.json')
+      );
+    } catch (e) {
+      if (!FileSystem.isNotExistError(e)) {
```
Comment (octogonz) [inline apps/heft/src/stages/CleanStage.ts:68]: A lot of Heft features are enabled based on the existence of their config file.  In the case where the file doesn't exist, we should be careful that `loadConfigurationFileAsync()` does not do any expensive work.

Also, if we could avoid throwing+catching tons of exceptions during normal operation, that would make the "Break on exceptions" debugger feature usable. :-)

**Review by octogonz — APPROVED** [review]

File: apps/heft/src/stages/CleanStage.ts [hunk 4]
```diff
@@ -36,16 +37,43 @@ export interface ICleanStageOptions {
...
     super(heftConfiguration, loggingManager, CleanStageHooks);
   }
 
-  protected getDefaultStageProperties(options: ICleanStageOptions): ICleanStageProperties {
+  protected async getDefaultStagePropertiesAsync(
+    options: ICleanStageOptions
+  ): Promise<ICleanStageProperties> {
+    const cleanConfigurationFileLoader: ConfigurationFile<ICleanConfigurationJson> = new ConfigurationFile<
+      ICleanConfigurationJson
+    >(path.resolve(__dirname, '..', 'schemas', 'clean.schema.json'), {
```
Comment (octogonz) [inline apps/heft/src/stages/CleanStage.ts:54]: I also still want to introduce the ".defaults.json" concept from PR https://github.com/microsoft/rushstack/pull/2054

Should we convert the `schemaPath` positional parameter to be part of the options, like below?
```ts
new ConfigurationFile<
        IPluginConfigurationJson
      >({
        jsonSchemaPath: path.join(__dirname, '..', 'schemas', 'clean.schema.json'), // <----
        jsonDefaultsPath: path.join(__dirname, '..', 'schemas', 'clean.defaults.json'), // <----
        propertyInheritanceTypes: { plugins: InheritanceType.append },
        jsonPathMetadata: {
          '$.plugins.*.plugin': {
            pathResolutionMethod: PathResolutionMethod.NodeResolve
          }
        }
      }
```

That way the strings won't get confused later when we introduce `jsonDefaultsPath` alongside `jsonSchemaPath`.

Comment (iclanton) [reply apps/heft/src/stages/CleanStage.ts:54]: Yeah I'm cool with that.

File: apps/heft/src/stages/CleanStage.ts (hunk 3 above)
Comment (iclanton) [reply apps/heft/src/stages/CleanStage.ts:58]: At some level, an exception will always be thrown, unless we use the `FileSystem.exists` API, which is less efficient.

**Pull request merged**